File: game-of-life/kotlin-arrow-game-of-life/src/main/kotlin/GameInteraction.kt
import Console.puts
import arrow.fx.IO

object GameInteraction {
    fun welcome(): IO<Unit> = puts("Welcome to the Conway's Game of Life!")

    fun readInitialMatrix(): IO<Matrix> =
            IO.just(parse(
                    this::class.java.classLoader
                            .getResource("4x4-block.csv")!!
                            .readText())
            )

    private fun parse(input: String): Matrix = input
            .split("\n")
            .map { line ->
                line.split(",").map { element ->
                    when (element) {
                        "Dead" -> Dead
                        else -> Alive
                    }
                }
            }

    fun print(matrix: Matrix): IO<Matrix> = puts(showMatrix.run { matrix.show() })
            .flatMap { IO.just(matrix) }
}

File: game-of-life/typescript-fat-arrow-ts-game-of-life/src/GameInteraction.ts
import {puts} from "./Console";
import {alive, dead, Matrix, showMatrix} from "./Data";

export const welcome: Promise<void> = puts("Welcome to the Conway's Game of Life!")

export const print: (matrix: Matrix) => Promise<Matrix> = (matrix: Matrix) => puts(showMatrix(matrix)).then(() => matrix)

export const readInitialMatrix: Promise<Matrix> = Promise.resolve([
    [dead(), alive(), dead()],
    [dead(), alive(), dead()],
    [dead(), alive(), dead()]
])
